### PR TITLE
Optimise WAL loading by caching min-time

### LINKED ISSUE #8638: Idea to speed up WAL loading
I observe that WAL loading can spend a lot of CPU time rejecting samples as already in chunks.
Would it be reasonable to cache `s.mmappedChunks[len(s.mmappedChunks)-1].maxTime` in `Head.processWALSamples()` , and thereby avoid calling and re-computing it?

```
  Total:      14.90s     15.31s (flat, cum) 32.93%
...
   2331         50ms       50ms           func (s *memSeries) append(t int64, v float64, appendID uint64, chunkDiskMapper *chunks.ChunkDiskMapper) (sampleInOrder, chunkCreated bool) { 
...
   2337            .       70ms           	c := s.head() 
   2338            .          .            
   2339        6.27s      6.27s           	if c == nil { 
   2340        4.22s      4.28s           		if len(s.mmappedChunks) > 0 && s.mmappedChunks[len(s.mmappedChunks)-1].maxTime >= t { 
   2341            .          .           			// Out of order sample. Sample timestamp is already in the mmaped chunks, so ignore it. 
   2342        4.31s      4.31s           			return false, false 
   2343            .          .           		} 
   2344            .          .           		// There is no chunk in this series yet, create the first chunk for the sample. 
   2345            .      120ms           		c = s.cutNewHeadChunk(t, chunkDiskMapper) 
   2346            .       10ms           		chunkCreated = true 
   2347            .          .           	} 
```

(This profile is from Cortex built with Prometheus code from commit c7e525bc6ba0ea6ef032f84279f73fae86d4143f)

### PR DESCRIPTION
**Pull request by bboreham**
Fixes #8638

Profiling shows that computing the last timestamp of mapped chunks inside `append()` takes ~10% of CPU time, so compute and store it once per series.

Edit: also fix some issues in the benchmark.

### CODE REVIEW

File: tsdb/head.go [hunk 1]
```diff
@@ -414,17 +418,24 @@ func (h *Head) processWALSamples(
...
-				ms = h.series.getByID(s.Ref)
-				if ms == nil {
+			if ms.ms == nil {
+				ms.ms = h.series.getByID(s.Ref)
+				if ms.ms == nil {
 					unknownRefs++
 					continue
 				}
+				if len(ms.ms.mmappedChunks) == 0 {
+					ms.minTime = -1
```
Comment (pracucci) [inline tsdb/head.go:428]: I'm not sure, but I think I've heard TSDB has been designed to work making no assumptions on the timestamp value. @codesome can share more light on this.

```suggestion
					ms.minTime = math.MinInt64
```

**Review by pracucci — COMMENTED** [review]
The change makes sense to me. What does `BenchmarkLoadWAL()` says (eg. running with `-count=5` and then comparing with `benchstat`?

File: tsdb/head.go (hunk 1 above)
Comment (roidelapluie) [reply tsdb/head.go:428]: 👍🏼

**Comment by bboreham** [comment]
As far as I can see `BenchmarkLoadWAL()` sets up a WAL with no mmapped chunks, so the code I noted in the profile is not executed.

I think I need some advice on how to set up a benchmark with both WAL samples and mmapped chunks.

**Comment by bboreham** [comment]
I spent some time drilling into the behaviour of the code, figured out how to create some mmapped chunks in `BenchmarkLoadWAL()`.
When I profile the benchmark, much of the time is spent in the Go scheduler; is that expected?
I think I found one flaw in the benchmark; I'm interested to hear whether it has been closely examined before.

**Comment by codesome** [comment]
> I'm interested to hear whether it has been closely examined before.

Prolly not

**Comment by bboreham** [comment]
I have added a new option to `BenchmarkLoadWAL()` to simulate mmapped chunks, and a few more commits to address some items noted at #9101.

Headline benchmark numbers say not much has changed:
```
name                                                                                                    old time/op    new time/op    delta
LoadWAL/batches=10,seriesPerBatch=100,samplesPerSeries=7200,exemplarsPerSeries=0,mmappedChunkT=0-4         582ms ±20%     537ms ± 7%    ~     (p=0.222 n=5+5)
LoadWAL/batches=10,seriesPerBatch=10000,samplesPerSeries=50,exemplarsPerSeries=0,mmappedChunkT=0-4         866ms ±17%     814ms ±10%    ~     (p=0.310 n=5+5)
LoadWAL/batches=10,seriesPerBatch=1000,samplesPerSeries=480,exemplarsPerSeries=0,mmappedChunkT=0-4         329ms ± 2%     326ms ± 5%    ~     (p=0.690 n=5+5)
LoadWAL/batches=100,seriesPerBatch=1000,samplesPerSeries=480,exemplarsPerSeries=0,mmappedChunkT=3800-4     3.26s ± 5%     3.21s ± 3%    ~     (p=0.421 n=5+5)

name                                                                                                    old alloc/op   new alloc/op   delta
LoadWAL/batches=10,seriesPerBatch=100,samplesPerSeries=7200,exemplarsPerSeries=0,mmappedChunkT=0-4        52.1MB ± 0%    52.1MB ± 0%  +0.07%  (p=0.008 n=5+5)
LoadWAL/batches=10,seriesPerBatch=10000,samplesPerSeries=50,exemplarsPerSeries=0,mmappedChunkT=0-4         260MB ± 8%     254MB ± 7%    ~     (p=0.310 n=5+5)
LoadWAL/batches=10,seriesPerBatch=1000,samplesPerSeries=480,exemplarsPerSeries=0,mmappedChunkT=0-4        51.5MB ± 2%    51.9MB ± 3%    ~     (p=0.841 n=5+5)
LoadWAL/batches=100,seriesPerBatch=1000,samplesPerSeries=480,exemplarsPerSeries=0,mmappedChunkT=3800-4     351MB ± 0%     361MB ± 1%  +2.69%  (p=0.008 n=5+5)

name                                                                                                    old allocs/op  new allocs/op  delta
LoadWAL/batches=10,seriesPerBatch=100,samplesPerSeries=7200,exemplarsPerSeries=0,mmappedChunkT=0-4          586k ± 0%      586k ± 0%    ~     (p=0.421 n=5+5)
LoadWAL/batches=10,seriesPerBatch=10000,samplesPerSeries=50,exemplarsPerSeries=0,mmappedChunkT=0-4         2.49M ± 0%     2.49M ± 0%    ~     (p=0.310 n=5+5)
LoadWAL/batches=10,seriesPerBatch=1000,samplesPerSeries=480,exemplarsPerSeries=0,mmappedChunkT=0-4          487k ± 0%      487k ± 0%    ~     (p=0.905 n=5+4)
LoadWAL/batches=100,seriesPerBatch=1000,samplesPerSeries=480,exemplarsPerSeries=0,mmappedChunkT=3800-4     3.73M ± 0%     3.73M ± 0%    ~     (p=0.841 n=5+5)
```

however if we look at the amount of CPU used during the test with simulated mmapped chunks we see about a 10% improvement, in line with what the profile showed at #8638:
```
before: 13.88user 2.82system 0:07.31elapsed 228%CPU (0avgtext+0avgdata 275464maxresident)k
after:  12.39user 2.26system 0:07.02elapsed 208%CPU (0avgtext+0avgdata 275956maxresident)k
```

**Review by cstyan — COMMENTED** [review]
Can we leave a comment somewhere to the effect that skipping samples that we know will be rejected in append is the optimization you're making around line 468/469. It's not immediately clear that caching of `minTime` is that important without looking at the linked issue.

**Comment by bboreham** [comment]
I started to re-word the comment, then realised the whole basis for this cache is incorrect.

"Mitigate lock contention in getByID", it said, but the locks are sharded 32,768 ways so with any likely number of workers there is no contention.

Removing `refSeries` altogether makes my test go about as fast as adding `minTime` did.  The real problem must be somewhere else.

**Comment by cstyan** [comment]
> Removing refSeries altogether makes my test go about as fast as adding minTime did. The real problem must be somewhere else.

Interesting.

**Comment by bboreham** [comment]
Replaced by #9160

**Pull request closed without merging**